### ISSUE TITLE
[Question] when is rendering trigger inside async flow

### ISSUE THREAD

**Issue by jiyinyiyong** [mobxjs/mobx-state-tree#519]
I run into a problem that I used a piece of missing data in my render method, which caused an error. 

![image](https://user-images.githubusercontent.com/449224/32761273-71e11fd0-c92d-11e7-9bbb-b72ff222d9d2.png)

I traced down to code and added some logs:

```ts
loadX: flow(function* loadX() {
  if (self.loadXState === RequestState.Processing) {
    return;
  }
  if (self.X) {
    return;
  }

  self.loadXState = RequestState.Processing;
  console.log("load X"); // <--- this one is printed
  try {
    let result = yield getXOfY(self.id);
    if (!self.X) {
      self.X = observable([]); // <--- probably the bug, I find `.X` not null so tried to find data, however no data yet
    }
    console.log("init X"); // <--- not printed yet
    applySnapshot(self.X, result.items);
    self.loadXState = RequestState.Succeeded;
  } catch (error) {
    self.loadXState = RequestState.Failed;
    throw error;
  }
}),
```

It appears that rendering is triggered as soon as `self.X=` happened. Was that the expected behavior? Is there other means rendering can be triggered?

**Comment by mweststrate** [comment]
That is correct, a flow breaks down a process in synchronous actions, and after each action the UI will be triggered to update if relevant data was updated. So your UI should be able to deal with that

```javascript
loadX: flow(function* loadX() {
  /* first action starts here */
  if (self.loadXState === RequestState.Processing) {
    return;
  }
  if (self.X) {
    return;
  }

  self.loadXState = RequestState.Processing;
  console.log("load X"); // <--- this one is printed
  try {
    let result = /* 1 action returns & ends here: */ yield getXOfY(self.id);
    /* action 2 */
    if (!self.X) {
      self.X = observable([]); // <--- probably the bug, I find `.X` not null so tried to find data, however no data yet
    }
    console.log("init X"); // <--- not printed yet
    applySnapshot(self.X, result.items);
    self.loadXState = RequestState.Succeeded;
  } catch (error) {
    self.loadXState = RequestState.Failed;
    throw error;
  }
}),
```

**Comment by jiyinyiyong** [comment]
I didn't make it clear enough. It took a break at `yield`, that's correct. But between `self.X = observable([])` and `/applySnapshot(self.X, result.items);` there's a rendering process, which I don't understand.

**Comment by mweststrate** [comment]
I am pretty sure that cannot be; JS is single threaded so it cannot leave the function at that point, except when an exception is thrown (but that should end up in your catch block). But the function cannot be interleaved by a render process at that point. 

Note that console.log itself prints async (although always in the execution order), so you might need to set breakpoints or use `window.alert` to be entirely sure when which code is executing

**Comment by jiyinyiyong** [comment]
will try tomorrow, sounds strange to me though

**Comment by chenyong** [comment]
Still me... I tried with with `alert`, the problem remains:

```js
self.loadXState = RequestState.Processing;
try {
  alert("before X load");
  let result = yield listXOfPlant(self.id);
  alert("network finished");
  if (!self.X) {
    alert("going to set");   //  <-- printed
    self.X = observable([]); //  <-- got a rendering event, which throws an error when it's []
    alert("after set");      //  <-- no printed
  }
  applySnapshot(self.X, result.items);
  alert("apply snapshot");
  self.loadXState = RequestState.Succeeded;
} catch (error) {
  self.loadXState = RequestState.Failed;
  throw error;
}
```

And the error:

```
X.tsx:33 Uncaught TypeError: Cannot read property 'processes' of undefined
    at X.render (X.tsx:33)
    at Object.allowStateChanges (mobx.module.js:969)
    at index.module.js:771
    at trackDerivedFunction (mobx.module.js:2908)
    at Reaction../node_modules/mobx/lib/mobx.module.js.Reaction.track (mobx.module.js:3076)
    at reactiveRender (index.module.js:766)
    at FlowProcesses.initialRender (index.module.js:759)
    at finishClassComponent (react-dom.development.js:10249)
    at updateClassComponent (react-dom.development.js:10226)
    at beginWork (react-dom.development.js:10605)
```

**Comment by d4rky-pl** [comment]
@chenyong Would it be possible for you to create a git repository that reproduces this problem?

**Comment by chenyong** [comment]
will try, but later...

**Comment by mweststrate** [comment]
Closing until repro is available

**Comment by zhangjunah** [comment]
I have the same problem。 And I create codesandbox: https://codesandbox.io/s/oqkzjm9wmy

when you click the 'load data' button, the console show like this:

![image](https://user-images.githubusercontent.com/1746154/33829460-4c94b0c8-deab-11e7-8523-5bee232e177a.png)

my model is like this:

![image](https://user-images.githubusercontent.com/1746154/33829512-8a23a02a-deab-11e7-9998-3cda38342f57.png)

**Comment by mweststrate** [comment]
Fixed in 1.3.1!
